extensions/amp-a4a/0.1/real-time-config-manager.js

```diff
@@ -65,331 +70,376 @@ export const RTC_ERROR_ENUM = {
   MACRO_EXPAND_TIMEOUT: '11',
 };
 
-/**
- * @param {string} error
- * @param {string} callout
- * @param {string} errorReportingUrl
- * @param {!Window} win
- * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampDoc
- * @param {number=} opt_rtcTime
- * @return {!Promise<!rtcResponseDef>}
- * @private
- */
-function buildErrorResponse_(
-  error, callout, errorReportingUrl, win, ampDoc, opt_rtcTime) {
-  dev().warn(TAG, `RTC callout to ${callout} caused ${error}`);
-  if (errorReportingUrl) {
-    sendErrorMessage(error, errorReportingUrl, win, ampDoc);
+export class RealTimeConfigManager {
+  /**
+   * @param {!./amp-a4a.AmpA4A} a4aElement
+   */
+  constructor(a4aElement) {
+    /** @private {!./amp-a4a.AmpA4A} */
+    this.a4aElement_ = a4aElement;
+
+    /** @private {?Object} */
+    this.urlReplacements_ = null;
+
+    /** @private {!Window} */
+    this.win_ = this.a4aElement_.win;
+
+    /** @private {!Object<string, boolean>} */
+    this.seenUrls_ = {};
+
+    /** @private {?number} */
+    this.rtcStartTime_ = null;
+
+    /** @private {!Array<!Promise<!rtcResponseDef>>} */
+    this.promiseArray_ = [];
+
+    /** @private {?RtcConfigDef} */
+    this.rtcConfig_ = null;
   }
-  return Promise.resolve(/**@type {rtcResponseDef} */(
-    {error, callout, rtcTime: opt_rtcTime || 0}));
-}
 
-/**
- * @param {string} errorType Uses the RTC_ERROR_ENUM above.
- * @param {string} errorReportingUrl
- * @param {!Window} win
- * @param {!../../../src/service/ampdoc-impl.AmpDoc} ampDoc
- */
-export function sendErrorMessage(errorType, errorReportingUrl, win, ampDoc) {
-  if (ERROR_REPORTING_ENABLED || getMode(win).localDev || getMode(win).test) {
-    const urls = Services.urlForDoc(ampDoc);
-    if (!urls.isSecure(errorReportingUrl)) {
-      dev().warn(TAG, `Insecure RTC errorReportingUrl: ${errorReportingUrl}`);
-      return;
-    }
-    const whitelist = {ERROR_TYPE: true, HREF: true};
-    const macros = {
-      ERROR_TYPE: errorType,
-      HREF: win.location.href,
-    };
-    const url = Services.urlReplacementsForDoc(ampDoc).expandUrlSync(
-        errorReportingUrl, macros, whitelist);
-    new win.Image().src = url;
+  /**
+   * Lazily initializes this.urlReplacements_.
+   * @return {!Object}
+   * @private
+   */
+  getUrlReplacements_() {
```

Comment:
my argument was that writing to a local field doesn't save you much as the service is already only loaded once.  Instead I would just reference when accessed (e.g. replace this.getUrlReplacements_() with Services.urlReplacementsForDoc(this.a4aElement_.getAmpDoc()) or you can store ampdoc as a field as change to just this.ampdoc)